atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/AsyncClusterSessionImpl.java

```diff
@@ -0,0 +1,135 @@
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.atlasdb.keyvalue.cassandra.async.AsyncSessionManager.CassandraClusterSessionPair;
+import com.palantir.common.concurrent.NamedThreadFactory;
+import com.palantir.common.concurrent.PTExecutors;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.tritium.metrics.registry.MetricName;
+
+public final class AsyncClusterSessionImpl implements AsyncClusterSession {
+
+    private static final Logger log = LoggerFactory.getLogger(AsyncClusterSessionImpl.class);
+
+    private final CassandraClusterSessionPair pair;
+    private final String sessionName;
+    private final Executor executor;
+
+    private final ScheduledExecutorService service = PTExecutors.newSingleThreadScheduledExecutor(
+            new NamedThreadFactory(sessionName() + "-healtcheck", true /* daemon */));
+    private final PreparedStatement healthCheckStatement;
+
+    public static AsyncClusterSessionImpl create(String clusterName, CassandraClusterSessionPair pair,
+            ThreadFactory threadFactory) {
+        return create(clusterName, pair, Executors.newCachedThreadPool(threadFactory));
+    }
+
+    public static AsyncClusterSessionImpl create(String clusterName, CassandraClusterSessionPair pair,
+            Executor executor) {
+        PreparedStatement healthCheckStatement = pair.session().prepare("SELECT dateof(now()) FROM system.local ;");
+        return new AsyncClusterSessionImpl(clusterName, pair, executor, healthCheckStatement);
+    }
+
+    private AsyncClusterSessionImpl(String sessionName, CassandraClusterSessionPair pair, Executor executor,
+            PreparedStatement healthCheckStatement) {
+        this.sessionName = sessionName;
+        this.pair = pair;
+        this.executor = executor;
+        this.healthCheckStatement = healthCheckStatement;
+    }
+
+    @Override
+    public Map<MetricName, Metric> usedCqlClusterMetrics() {
+        return KeyedStream.stream(
+                pair.cluster().getMetrics()
+                        .getRegistry()
+                        .getMetrics())
+                .mapKeys(name -> MetricName.builder().safeName(name).build())
+                .collectToMap();
+    }
+
+    @Nonnull
+    @Override
+    public String sessionName() {
+        return sessionName;
+    }
+
+    @Override
+    public void close() {
+        service.shutdownNow();
+        log.info("Shutting down health checker for cluster session {}", SafeArg.of("clusterSession", sessionName));
+        boolean shutdown = false;
+        try {
+            shutdown = service.awaitTermination(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            log.error("Interrupted while shutting down health checker. Should not happen");
+            Thread.currentThread().interrupt();
+        } finally {
+            AsyncSessionManager.getOrInitializeAsyncSessionManager().closeClusterSession(this);
```

Comment:
I am using this `close` method on `AsyncClusterSessionManager` as there were several ideas on how to close sessions when they are not used anymore. One of the options was to do reference counting from it. Because of that what in fact happens is that we register that this session is not used by a C*KVS anymore. Logic definitely has to change and naming should be more meaningful.

Reply:
I changed the `start` and `close` logic, we should discuss if it can be done in a more logical manner.